kunlar/12-kun-Regular-ifodalar/12-kun-regular-ifodalar.md

```diff
@@ -0,0 +1,412 @@
+        print(qator)
+```
+
+`"^From:.+@"` ni tarjima qilsak
+> (^)So'z boshidan `From:` va undan keyin birdan ortiq belgilar kelsin 
+> va undan keyin @ (kuchukcha) belgisi kelsin
+
+Konsolga quyidagi chiqadi:
+
+```commandline
+From: stephen.marquard@uct.ac.za
+From: louis@media.berkeley.edu
+From: zqian@umich.edu
+From: rjlowe@iupui.edu
+From: zqian@umich.edu
+From: rjlowe@iupui.edu
+From: cwen@iupui.edu
+From: cwen@iupui.edu
+From: gsilver@umich.edu
+From: gsilver@umich.edu
+From: zqian@umich.edu
+From: gsilver@umich.edu
+From: wagnermr@iupui.edu
+From: zqian@umich.edu
+From: antranig@caret.cam.ac.uk
+From: gopal.ramasammycook@gmail.com
+From: david.horwitz@uct.ac.za
+From: david.horwitz@uct.ac.za
+From: david.horwitz@uct.ac.za
+From: david.horwitz@uct.ac.za
+From: stephen.marquard@uct.ac.za
+From: louis@media.berkeley.edu
+From: louis@media.berkeley.edu
+From: ray@media.berkeley.edu
+From: cwen@iupui.edu
+From: cwen@iupui.edu
+From: cwen@iupui.edu
+```
+
+Natijadan ko'rishimiz mumkinki hamma qator `From:` dan boshlangan va `@` belgisi bor.
+
+
+## `findall` va `\S`
+
+Shu paytgacha biz `re.search` funksiyasidan foydalandik. Endi `findall` funksiyasidan foydalanamiz
+hamda `\S` operatoridan foydalanishni ko'ramiz. Bu funksiya berilgan kalit so'z orqali tekstdan qidiradi
+va topilganlarni `list` shaklida qaytaradi. Masalan, quyidagi tekstni olaylik
+```python
+# \S+@\S+: berilgan stringdan shunday substringlarni qidiringki, undagi substringlar boshlanish belgilardan bulib
+# undan keyin @ belgi bo'lib va undan keyin yana belgilar bo'lishi kerak
+s = 'Biz xabarni ulugbek@gmail.com dan oldik, keyin uni nuriddin@gmail.com ga yubordik'
+lst = re.findall('\S+@\S+', s)
+print(lst)
+```
+
+`re.findall` ga `'\S+@\S+'` maxsus belgilarni joyladik. `'\S+@\S+'` degani, shunday so'zlarni teksdan qidiringki
+- `\S+` ularni boshi belgilardan tashkil topsin, lekin belgilar birortasi bo'sh belgi (`" "`) bo'lmaligi kerak 
+- `@` undan keyin kuchukcha belgisi bo'lsin
+- `\S+` ularni oxiri belgilardan tashkil topsin, lekin belgilar birortasi bo'sh belgi (`" "`) bo'lmaligi kerak
+
+Ishlatsak konsolga quyidagicha natija chiqadi:
+
+```commandline
+['ulugbek@gmail.com', 'nuriddin@gmail.com']
+```
+
+Xuddi shu maxsus belgilarni faylga qo'llasak, quyidagicha bo'ladi:
+
+```python
+fayl = open("regifoda.txt", 'r')
+for qator in fayl.readlines():
+    qator = qator.rstrip()
+    lst = re.findall('\S+@\S+', qator)
+    if lst != []:
+        print(lst)
+```
+
+Konsolga
+
+```commandline
+['stephen.marquard@uct.ac.za']
+['<postmaster@collab.sakaiproject.org>']
+['<200801051412.m05ECIaH010327@nakamura.uits.iupui.edu>']
+['<source@collab.sakaiproject.org>;']
+['<source@collab.sakaiproject.org>;']
+['<source@collab.sakaiproject.org>;']
+['apache@localhost)']
+```
+
+Yuqoridagi natija elektron adreslarni chiqarishdir. Lekin, etibor bersak
+bazilarida xar xil biz xoxlamaydigan belgilar mavjud xususan `<;>`. 
+
+> Qanday qilib, ushbu belgilarni olish tashlaymiz? 
```

Comment:
olish tashlaymiz - olib tashlaymiz boladi

Reply:
To'g'irlandi. Rahmat.